File: benchmarks/cross-language/elixir/fork_join.exs
# Elixir Fork-Join Benchmark (Savina-style)

defmodule ForkJoin do
  @num_workers 8
  @messages_per_worker (System.get_env("BENCHMARK_MESSAGES") |> case do
    nil -> 100_000
    val -> div(String.to_integer(val), 8)
  end)

  def start do
    total = @num_workers * @messages_per_worker
    IO.puts("=== Elixir Fork-Join Throughput Benchmark ===")
    IO.puts("Workers: #{@num_workers}, Messages: #{total}\n")

    parent = self()

    # Create workers
    workers = Enum.map(1..@num_workers, fn _ ->
      spawn(fn -> worker_loop(0, parent) end)
    end)
    worker_tuple = List.to_tuple(workers)

    start_time = :erlang.monotonic_time(:nanosecond)

    # Send messages round-robin
    send_work(worker_tuple, 0, total)

    # Signal workers to report
    Enum.each(workers, fn w -> send(w, {:done, self()}) end)

    # Collect results
    total_processed = collect_results(@num_workers, 0)

    end_time = :erlang.monotonic_time(:nanosecond)
    elapsed_ns = end_time - start_time
    elapsed_sec = elapsed_ns / 1_000_000_000.0

    if total_processed != total do
      IO.puts("VALIDATION FAILED: expected #{total}, got #{total_processed}")
    end

    throughput = total / elapsed_sec / 1_000_000
    ns_per_msg = elapsed_ns / total

    IO.puts("ns/msg:         #{:io_lib.format("~.2f", [ns_per_msg])}")
    IO.puts("Throughput:     #{:io_lib.format("~.2f", [throughput])} M msg/sec")
  end

  defp send_work(_workers, i, total) when i >= total, do: :ok
  defp send_work(workers, i, total) do
    worker_idx = rem(i, @num_workers)
    send(elem(workers, worker_idx), {:work, i})
    send_work(workers, i + 1, total)
  end

  defp worker_loop(processed, parent) do
    receive do
      {:work, _value} ->
        worker_loop(processed + 1, parent)
      {:done, from} ->
        send(from, {:result, processed})
    end
  end

  defp collect_results(0, acc), do: acc
  defp collect_results(n, acc) do
    receive do
      {:result, count} ->
        collect_results(n - 1, acc + count)
    end
  end
end

ForkJoin.start()

File: benchmarks/cross-language/erlang/fork_join.erl
-module(fork_join).
-export([start/0]).

-define(NUM_WORKERS, 8).

get_messages_per_worker() ->
    Total = case os:getenv("BENCHMARK_MESSAGES") of
        false -> 100000;
        Val -> list_to_integer(Val)
    end,
    Total div ?NUM_WORKERS.

start() ->
    MessagesPerWorker = get_messages_per_worker(),
    io:format("=== Erlang Fork-Join Throughput Benchmark ===~n"),
    Total = ?NUM_WORKERS * MessagesPerWorker,
    io:format("Workers: ~p, Messages: ~p~n~n", [?NUM_WORKERS, Total]),

    Parent = self(),

    % Create workers
    Workers = [spawn(fun() -> worker(0, 0, Parent) end) || _ <- lists:seq(1, ?NUM_WORKERS)],
    WorkerArray = list_to_tuple(Workers),

    Start = erlang:monotonic_time(nanosecond),

    % Send messages round-robin
    send_work(WorkerArray, 0, Total),

    % Signal workers to report
    [W ! {done, self()} || W <- Workers],

    % Collect results
    TotalProcessed = collect_results(?NUM_WORKERS, 0),

    End = erlang:monotonic_time(nanosecond),
    ElapsedNs = End - Start,
    ElapsedSec = ElapsedNs / 1000000000.0,

    if
        TotalProcessed =/= Total ->
            io:format("VALIDATION FAILED: expected ~p, got ~p~n", [Total, TotalProcessed]);
        true -> ok
    end,

    NsPerMsg = ElapsedNs / Total,
    MsgPerSec = Total / ElapsedSec,
    io:format("ns/msg:         ~.2f~n", [NsPerMsg]),
    io:format("Throughput:     ~.2f M msg/sec~n", [MsgPerSec / 1000000]),

    halt(0).

send_work(_Workers, I, Total) when I >= Total -> ok;
send_work(Workers, I, Total) ->
    WorkerIdx = (I rem ?NUM_WORKERS) + 1,
    element(WorkerIdx, Workers) ! {work, I},
    send_work(Workers, I + 1, Total).

worker(Processed, Sum, Parent) ->
    receive
        {work, Value} ->
            worker(Processed + 1, Sum + Value, Parent);
        {done, From} ->
            From ! {result, Processed}
    end.

collect_results(0, Acc) -> Acc;
collect_results(N, Acc) ->
    receive
        {result, Count} ->
            collect_results(N - 1, Acc + Count)
    end.
